sklearn/cluster/optics_.py

```diff
@@ -527,3 +568,382 @@ def cluster_optics_dbscan(reachability, core_distances, ordering, eps=0.5):
+
+
+class _Area:
+    """An (upward or downward) area.
+
+    Attributes
+    ----------
+    start : integer
+        The start of the region.
+
+    end : integer
+        The end of the region.
+
+    maximum : float
+        The maximum reachability in this region, which is the
+        start of the region for a downward area and the end of
+        the region for an upward area.
+
+    mib : float
+        Maximum in between value, i.e. the maximum value between
+        the end of a steep down area and the current index. It is
+        irrelevant for steep up areas.
+    """
+    def __init__(self, start, end, maximum, mib):
+        self.start = start
+        self.end = end
+        # maximum is the beginning of a downward area
+        # and the end of an upward area
+        self.maximum = maximum
+        # maximum in between (ref Sec. 4.3.2)
+        self.mib = mib
+
+    def __repr__(self):
+        return ("start: %s, end: %s, max: %4g, mib: %4g" %
+                (self.start, self.end, self.maximum, self.mib))
+
+
+def _extend_region(steep_point, xward_point, start, min_samples):
+    """Extend the area until it's maximal.
+
+    It's the same function for both upward and downward reagions, depending on
+    the given input parameters. To extend an upward reagion,
+    ``steep_point=steep_upward`` and ``xward_point=downward`` are expected, and
+    to extend a downward region, ``steep_point=steep_downward`` and
+    ``xward_point=upward``.
+
+    Parameters
+    ----------
+    steep_point : bool array, shape (n_samples)
+        True if the point is steep downward (upward).
+
+    xward_point : bool array, shape (n_samples)
+        True if the point is an upward (respectively downward) point.
+
+    start : integer
+        The start of the xward region.
+
+    min_samples : integer
+       The same as the min_samples given to OPTICS. Up and down steep
+       regions can't have more then ``min_samples`` consecutive non-steep
+       points.
+
+    Returns
+    -------
+    index : integer
+        The current index iterating over all the samples, i.e. where we are up
+        to in our search.
+
+    end : integer
+        The end of the region, which can be behind the index. The region
+        includes the ``end`` index.
+    """
+    n_samples = len(steep_point)
+    non_xward_points = 0
+    index = start
+    end = start
+    # find a maximal area
+    while index < n_samples:
+        if steep_point[index]:
+            non_xward_points = 0
+            end = index
+        elif not xward_point[index]:
+            # it's not a steep point, but still goes up.
+            non_xward_points += 1
+            # region should include no more than min_samples consecutive
+            # non steep xward points.
+            if non_xward_points > min_samples:
+                break
+        else:
+            return min(index, n_samples - 1), end
+
+        index += 1
+    return min(index + 1, n_samples - 1), end
```

Comment:
``end`` seems redundant, the extra points you go through can't be steep_upward points or steep_downward points?

Reply:
`end` is not redundant since it can be on the left side of `index` when the function returns. The loop continues for `downward`/`upward` points, even if they're not _steep_.

Reply:
I can't understand your point, a steep area has to been maximum?

Reply:
could you please rephrase the question?

Reply:
As far as I'm concerned, `end` is the main think this function is trying to calculate. `index` is returned merely to update the index at which we are searching, and may only be an efficiency bonus to avoid duplicated work. The one cannot be derived directly from the other.

Reply:
The next index to explore is only for efficiency, I think. It is certainly not explicit from the algorithm in the paper. It is just a convenience to, as in the description, indicate "where we are up to in our search."